lib/pp.rb

```diff
@@ -0,0 +1,657 @@
+#
+# It returns <tt>obj(s)</tt>.
+#
+##
+# == Output Customization
+#
+# To define a customized pretty printing function for your classes,
+# redefine method <code>#pretty_print(pp)</code> in the class.
+#
+# <code>#pretty_print</code> takes the +pp+ argument, which is an instance of the PP class.
+# The method uses #text, #breakable, #nest, #group and #pp to print the
+# object.
+#
+##
+# == Pretty-Print JSON
+#
+# To pretty-print JSON refer to JSON#pretty_generate.
+#
+##
+# == Author
+# Tanaka Akira <akr@fsij.org>
+
+class PP < PrettyPrint
+  # Returns the usable width for +out+.
+  # As the width of +out+:
+  # 1. If +out+ is assigned to a tty device, its width is used.
+  # 2. Otherwise, or it could not get the value, the +COLUMN+
+  #    environment variable is assumed to be set to the width.
+  # 3. If +COLUMN+ is not set to a non-zero number, 80 is assumed.
+  #
+  # And finally, returns the above width value - 1.
+  # * This -1 is for Windows command prompt, which moves the cursor to
+  #   the next line if it reaches the last column.
+  def PP.width_for(out)
+    begin
+      require 'io/console'
+      _, width = out.winsize
+    rescue LoadError, NoMethodError, SystemCallError
+    end
+    (width || ENV['COLUMNS']&.to_i&.nonzero? || 80) - 1
+  end
+
+  # Outputs +obj+ to +out+ in pretty printed format of
+  # +width+ columns in width.
+  #
+  # If +out+ is omitted, <code>$></code> is assumed.
+  # If +width+ is omitted, the width of +out+ is assumed (see
+  # width_for).
+  #
+  # PP.pp returns +out+.
+  def PP.pp(obj, out=$>, width=width_for(out))
+    q = PP.new(out, width)
+    q.guard_inspect_key {q.pp obj}
+    q.flush
+    #$pp = q
+    out << "\n"
+  end
+
+  # Outputs +obj+ to +out+ like PP.pp but with no indent and
+  # newline.
+  #
+  # PP.singleline_pp returns +out+.
+  def PP.singleline_pp(obj, out=$>)
+    q = SingleLine.new(out)
+    q.guard_inspect_key {q.pp obj}
+    q.flush
+    out
+  end
+
+  # :stopdoc:
+  def PP.mcall(obj, mod, meth, *args, &block)
+    mod.instance_method(meth).bind_call(obj, *args, &block)
+  end
+  # :startdoc:
+
+  if defined? ::Ractor
+    class << self
+      # Returns the sharing detection flag as a boolean value.
+      # It is false (nil) by default.
+      def sharing_detection
+        Ractor.current[:pp_sharing_detection]
+      end
+      # Sets the sharing detection flag to b.
+      def sharing_detection=(b)
+        Ractor.current[:pp_sharing_detection] = b
+      end
+    end
+  else
+    @sharing_detection = false
+    class << self
+      # Returns the sharing detection flag as a boolean value.
+      # It is false by default.
+      attr_accessor :sharing_detection
+    end
+  end
+
+  module PPMethods
+
+    # Yields to a block
+    # and preserves the previous set of objects being printed.
+    def guard_inspect_key
+      if Thread.current[:__recursive_key__] == nil
+        Thread.current[:__recursive_key__] = {}.compare_by_identity
```

Comment:
It looks like we only use the Thread class for thread-local storage. I think we can replace this with `Fiber.current` to get the same functionality, which saves us from the bazillion other changes related to Thread in the specs.

Reply:
I'm ok with the changes. I think it's a milestone to copy PrettyPrinter as-is without any tweaks needed.